Add smarter type inference for ofType pipe.

rkirov opened the pull request:
It is based on TS2.8 introduced conditional types. Upgrade package.json
to that version.

Also remove the deprecated non-pipable version of `ofType` so that we
don't have to keep complicated types in two places.

Fixes some tests post removal of non-pipable version.

Original implementation by @mtaran-google.

Fixes #860

rkirov commented on modules/effects/compat/rollup.config.js:
```diff
@@ -0,0 +1,12 @@
+export default {
```
I copied those from `testing` but I have not idea whether they are correct in the context of compat.

timdeschryver commented:
I think we should allow an "infite" number of `ofType` arguments?

mtaran-google commented:
That would be ideal, but the type system in TS 2.9 isn't powerful enough to check that many. 3.0 (in RC now) [improves on this](https://blogs.msdn.microsoft.com/typescript/2018/07/12/announcing-typescript-3-0-rc/#tuples-and-parameters) but I'm not sure if it'll be enough. It'll be interesting to check out when it lands.

timdeschryver commented:
@mtaran-google I see, thanks.

Just one more think, I thought the following would give me compile errors but they don't:
```ts
// this is OK
ofType<Search>(BookActionTypes.Search),

// I would expect an error here
ofType<Search>('sss'),

// I would expect an error here, because the payload of `SearchComplete` is an array
ofType<Search>(BookActionTypes.SearchComplete),
map(action => action.payload),
  switchMap(query => {
    if (query === '') {
        return empty();
     }
```

I fiddled with it in the past and came up with [this](https://github.com/timdeschryver/platform/commit/6cca61c08a1aa23b999bdaa0756965c9f15be794) which is giving compile errors.

rkirov commented:
@timdeschryver That's a good point. Let's recap the desired outcomes:

1) ofType('x') to just work based in inference
2) ofType<SomeType>('someActionType') to keep working as is.
3) ofType<SomeType>('typo') to error if SomeType.type != 'typo' (both of your examples boil down to this).

My current PR has 2) but not 3), and yours I think has 3) but not 2), because you have two free type variables "export function ofType<V extends Action, Type extends string>"  and no defaults. Maybe I am missing something, but I don't think you can have 2) and 3) at the same time.

rkirov commented:
However, based on your your example it looks like the T1,T2,..., etc expansions are not necessary. I think the following works the same.

```typescript
export function ofType<
  V extends Extract<U, { type: T }>,
  T extends string = string,
  U extends Action = Action
>(...t: T[]): OperatorFunction<U, V>;
```

Also one final thing, my work here is prompted by attempting to turn on --strictFunctionTypes and hitting a lot of ngrx issues. Please do your testing with that flag on. I think any solution that doesn't have at least two parameters U -> V to capture the broader observable input and the narrowed out observable output, will not work with --strictFunctionTypes.

alex-okrushko commented:
`Extract` showed up in 2.8, while ngrx is still at 2.7

rkirov commented:
@alex-okrushko That is the blocker for all this. TS2.8 is required for conditional types, Maksym's original change a custom alias for `Extract`.

timdeschryver commented:
@rkirov Yep you're completely right.
The solution I came up with doesn't support `ofType<SomeType>('someActionType')`.

And the snippet you posted has the same outcome.
```ts
export function ofType<
  V extends Extract<U, { type: T }>,
  T extends string = string,
  U extends Action = Action
>(...t: T[]): OperatorFunction<U, V>;
```

rkirov commented:
Ok, I just discovered one more issue that might be the deal breaker here. A lot of teams do not pass they type argument on their actions observable is it just of type - `Actions`. Now, if that property is `Actions<AllPossibleActions>` this change is great.

However, for unqualifed `Actions` this change (in both proposed forms) is actually a regression. That means if one has `private actions: Actions;` then `this.actions.pipe(ofType('something'))` creates `Observable<never>` instead of at least `Observable<Action>`.

This kinda expected because `Extract<{type: string}, {type: 'a'}>` is `never` and not `{type: 'a'}`. In a more general form `Extract<string, 'a'>` is `never` though one can expect it to be 'a'. Not sure if TS has plans to change that.

rkirov commented:
I updated this PR with what is my final proposal for the future types of 'ofType'. Added comments that should capture the upside and downsides of that approach. Hopefully, that would make the type soup a bit more maintainable.

Internally, I appears we have an agreement  that this is good and are going to change it with a patch before this PR lands here. Here it appears to be stuck behind angular, bazel, typescript versioning hell, that I can't resolve.

@MikeRyanDev and @brandonroberts please at least verify that there are no major concerns with this proposal, otherwise we would end up forking out internal google users from the mainstream branch here.

MikeRyanDev commented:
@rkirov This proposed change looks good to me.

timdeschryver commented on modules/effects/src/actions.ts:
```diff
@@ -1,8 +1,11 @@
 import { filter } from 'rxjs/operators';
 
+// Removes types from T that are not assignable to U.
+export type Filter<T, U> = T extends U ? T : never;
```
Because we're using `Extract` this can be removed now.

rkirov replied:
Indeed.

timdeschryver commented on modules/effects/spec/actions.spec.ts:
```diff
@@ -58,10 +66,10 @@ describe('Actions', function() {
+  it('should let you filter out actions', function() {
     actions$
       .pipe(
         ofType(ADD),
```
TS is complaining about the next line `map(update => update.type)`
This can be solved if we define our actions as `let actions$: Actions<AddAcction | SubtractAction>;`

rkirov replied:
Sorry, I haven't even been able to run the typechecker on the whole PR, because of the other deps. Fixed.

rkirov replied:
CI is green. I assume this means it is not longer an issue.

timdeschryver replied:
LGTM 👍

timdeschryver commented on modules/effects/src/actions.ts:
```diff
@@ -19,17 +22,79 @@ export class Actions<V = Action> extends Observable<V> {
     return observable;
   }
-
-  ofType<V2 extends V = V>(...allowedTypes: string[]): Actions<V2> {
```
This may sound stupid, but here it goes anyways:
Since we have the fallback `ofType`, could we not remove this but mark it as deprecated?
With the next release we can remove it.
I say this because while the `compat` solution works, the user would have to change the import statement?

rkirov replied:
I have no opinions on backwards compat (google operates on one-version policy), and will do whatever you guys decide.

brandonroberts replied:
I was planning on doing a 6.1 release, so maybe we can deprecate the instance operators before release and then move forward with the removal for the next major release.

alex-okrushko commented on package.json:
```diff
@@ -143,7 +143,7 @@
     "tslint": "^5.0.0",
     "tsutils": "2.20.0",
-    "typescript": "~2.7.2",
+    "typescript": "~2.8.4",
```
I think this means that angular version would need a bump as well:
```
@angular/compiler-cli@6.0.9 requires typescript@'>=2.7.0 <2.8.0' but 2.8.4 was found instead.
```

rkirov replied:
Indeed, at this point this PR is doing too much. Another PR is needed to get to TS 2.8, with all the complications involved around angular, bazel, router(?). I can't do that upgrade, but will rebase this PR once that happens.

rkirov commented:
Opened https://github.com/ngrx/platform/issues/1220 to track the compiler upgrade. This PR is blocked (will respond to review comments though) until that gets resolved.

rkirov commented:
rebased the PR, will take a look at the CI results tomorrow.

rkirov commented:
CI is green, please proceed with a regular review. I don't fully understand the compat strategy, mostly copy/pasted from other directories.

timdeschryver commented:
I think the `compat` version won't be needed. The `store.ofType` (and `store.select`) operators are now marked as deprecated and will be removed the next version.
Before you change anything, wait untill @brandonroberts 's confirmation tho.

brandonroberts commented:
@rkirov @timdeschryver I think its safe to remove the `compat` package also. The commit message does need to reflect the breaking change of removing the instance operator.

alex-okrushko commented:
@rkirov Hey Rado. Would you be able to take a look at this PR again?

dimavt commented:
Any update on this? I can finalize this PR, if you don't mind.

For someone who wants to try this feature in your project today, you can add this to your codebase:
```ts
declare module '@ngrx/effects' {
  export function ofType<
    V extends Extract<U, { type: T }>, 
    T extends string = string, 
    U extends Action = Action
  >(...actions: T[]): OperatorFunction<U, V>;
}
```

timdeschryver commented:
If you want I can finish this one up @rkirov but you'll have to give maintainers write access to your branch.

rkirov commented:
I will rebase it today. Sorry for the delay.

Btw, we have been running this with change internally in google for some time and I haven't heard any complaints.

rkirov commented on modules/effects/spec/actions.spec.ts:
```diff
@@ -11,12 +11,20 @@ import { hot, cold } from 'jasmine-marbles';
 
 describe('Actions', function() {
-  let actions$: Actions;
+  let actions$: Actions<AddAcction | SubtractAction>;
```
done

rkirov commented:
Rebased, removed compat bits and updated commit message with 'BREAKING CHANGE'. Please take another look.

brandonroberts commented:
Thanks @rkirov!
